File: python/graph.py
#!/usr/bin/env python3

from collections import deque

class GraphNode():
    def __init__(self, name, data=None):
        self.data = data
        self.name = name
        # We use a set here because we want to efficiently add,
        # remove, and iterate over all edges.
        self.edges = set()

        # We also maintain a set of backward edges from dest->source
        # so we can go backwards even though it is a directed graph.
        self.back_edges = set()
        self.color = "none-set"
        self.previous = None
        
    def connect(self, dest_node, weight=1):
        """Connects this node to another node.  Important notes: It
        assumes a DAG, and you need to connect the other way to be
        sure.  It also does NOT allow multiple connections to the same
        destination, and enforces this with an assertion.
        """
        for e in self.edges:
            if e.dest_node == dest_node:
                raise Exception("An edge already exists")
        edge = GraphEdge(self, dest_node, weight)
        self.edges.add(edge)
        dest_node.back_edges.add(edge)

    def connected(self, dest_node):
        for e in self.edges:
            if e.dest_node == dest_node:
                return True
        return False
        
    def disconnect(self, dest_node):
        edge = None
        for e in self.edges:
            if e.dest_node == dest_node:
                edge = e
        if edge == None:
            raise Exception("No edge exists between {} and {}".format(
                self.name, dest_node.name))
        self.edges.remove(edge)
        dest_node.back_edges.remove(edge)

    def __bool__(self):
        """Needed to keep python from getting too smart because we
        may also have __len__ defined"""
        return True

    def _check_structure(self):
        """A sanity check function that makes sure that every edge is well
        formed"""
        for edge in self.edges:
            assert edge in edge.dest_node.back_edges
            assert edge.source_node == self
        for edge in self.back_edges:
            assert edge in edge.source_node.edges
            assert edge.dest_node == self
        return True
    
    def __repr__(self):
        return ("( {}/{} [".format(self.name, self.data) +
                ", ".join(
                    map(
                        lambda x: "{}->{}".format(x.weight,
                                                  x.dest_node.name), self.edges))
                    + "])")

    def all_edges(self):
        """This iterates over all nodes it is connected TO in
        unspecified order, returning a tuple of destination/weight."""
        if not self.edges:
            return iter([])
        for edge in self.edges:
            yield (edge.dest_node, edge.weight)
        
class GraphEdge():
    def __init__(self, source_node, dest_node, weight=1):
        self.source_node = source_node
        self.dest_node = dest_node
        self.weight = weight

class Graph():
    def __init__(self, dictionary=None):
        self.nodes = {}
        if dictionary:
            for key in dictionary:
                self[key] = dictionary[value]

        
    def __getitem__(self, name):
        """This returns the NODE itself, not the data in the node."""
        if name not in self.nodes:
            raise IndexError("Unable to find {}".format(name))
        return self.nodes[name]

    def __setitem__(self, name, data):
        if name in self.nodes:
            self.nodes[name].data = data
        else:
            self.nodes[name] = GraphNode(name, data)

    def __delitem__(self, name):
        if name not in self.nodes:
            raise IndexError("Unable to find {}".format(name))
        node = self.nodes[name]
        for edge in node.edges:
            edge.dest_node.back_edges.remove(edge)
        for edge in node.back_edges:
            edge.source_node.edges.remove(edge)
        del self.nodes[name]

    def __contains__(self, name):
        return name in self.nodes
        
    def connect(self, source_name, dest_name, weight=1):
        source = self[source_name]
        dest = self[dest_name]
        source.connect(dest, weight)

    def disconnect(self, source_name, dest_name):
        source = self[source_name]
        dest = self[dest_name]
        source.disconnect(dest)
        
    def connected(self, source_name, dest_name):
        source = self[source_name]
        dest = self[dest_name]
        return source.connected(dest)
        
    def check_structure(self):
        """Will raise an assertion failure if it is not well formed"""
        for node in self.nodes:
            self.nodes[node]._check_structure()
        return True


    def dijkstra_traversal(self, node_name):
        start = self.nodes[node_name]
        unvisited = set()
        for n in self.nodes:
            self.nodes[n].distance = float('inf')
            self.nodes[n].previous = None
            unvisited.add(self.nodes[n])
            for (dest, weight) in self.nodes[n].all_edges():
                assert weight > 0
        start.distance = 0
        while len(unvisited) > 0:
            min_node = None
            for node in unvisited:
                if min_node == None:
                    min_node = node
                if node.distance < min_node.distance:
                    min_node = node
            if min_node.distance == float('inf'):
                break
            unvisited.remove(min_node)
            for (dest, weight) in min_node.all_edges():
                if ( dest in unvisited and
                     min_node.distance + weight < dest.distance):
                    dest.distance = min_node.distance + weight
                    dest.previous = min_node
            yield min_node

    def connected_from(self, node_name):
        for item in self.bfs_traversal(node_name):
            pass
        for n in self.nodes:
            if self.nodes[n].color == "white":
                return False
        return True
            
    def bfs_traversal(self, node_name):
        """Does an iterative breadth first search traversal"""
        for n in self.nodes:
            self.nodes[n].color = "white"
            self.nodes[n].previous = None
        start_node = self[node_name]
        queue = deque()
        start_node.color = "gray"
        queue.appendleft(start_node)
        while len(queue) != 0:
            node = queue.pop()
            for (dest, weight) in node.all_edges():
                if dest.color == "white":
                    dest.color = "gray"
                    dest.previous = node
                    queue.appendleft(dest)
            node.color = "black"
            yield node


    def dfs_iterative_traversal(self, node_name):
        """Does an iterative breadth first search traversal"""
        for n in self.nodes:
            self.nodes[n].color = "white"
            self.nodes[n].previous = None
        start_node = self[node_name]
        queue = []
        queue.append(start_node)
        start_node.color = "gray"
        while len(queue) != 0:
            node = queue[len(queue)-1]
            appended = False
            for (dest, weight) in node.all_edges():
                if dest.color == "white":
                    dest.color = "gray"
                    dest.previous = node
                    queue.append(dest)
                    appended = True
            if not appended:
                node = queue.pop()
                node.color = "black"
                yield node

            
    def dfs_traversal(self, node_name):
        def dfs_internal(at):
            at.color = "gray"
            for (dest, edge) in at.all_edges():
                if dest.color == "white":
                    dest.previous = at
                    yield from dfs_internal(dest)
            at.color = "black"
            yield at
        for n in self.nodes:
            self.nodes[n].color = "white"
            self.nodes[n].previous = None
        yield from dfs_internal(self.nodes[node_name])
        
    def __len__(self):
        return len(self.nodes)

    def __repr__(self):
    
        return "{ " + ", ".join(map(lambda x: repr(self.nodes[x]),
                                    self.nodes)) + "}" 

def make_star():
    g = Graph()
    nums = "01234"
    for i in range(5):
        g[i] = "Node{}".format(i)
    for i in range(5):
        g.connect(i, (i+2) % 5)
        g.connect(i, (i-2) % 5)
    return g
        

def samplegraph():
    g = Graph()
    connections = [(0,1,1), (0,5,5),
                   (1,2,20), (1,3,3), (1,4,4),
                   (2,0,2), (3,4,4),
                   (4,2,4)]
    for i in range(6):
        g[i] = "Node{}".format(i)
    for source, dest, weight in connections:
        g.connect(source, dest, weight)
    return g

File: kotlin/src/main/kotlin/Graph.kt
package edu.ucdavis.cs.nweaver

class GraphKeyError(message: String) : Exception(message)
class GraphEdgeError(message: String) : Exception(message)

/*
 * This is an adjacency list representation:  Every node has a list
 * of edges, and every edge has a reference to both the starting node
 * and the ending node for the edge.
 *
 * This is designed to represent a directed graph.  The value of the
 * node is designed to be an arbitrary type.
 */
class GraphNode<T>(val name: T){
    val outEdges: MutableSet<GraphEdge<T>> = mutableSetOf()
    val inEdges: MutableSet<GraphEdge<T>> = mutableSetOf()

    /*
     * The first thing we do with our classes is make sure we
     * have a meaningful toString function, because this greatly
     * eases the debugging process.
     */
    override fun toString(): String {
        return "{Node $name}"
    }
}

/*
 * And the class for edges.  Our edges are directional, from "start" to "end",
 * with a positive weight.
 */
class GraphEdge<T>(
    val start: GraphNode<T>,
    val end: GraphNode<T>,
    val weight: Double = 1.0 ) {
    init {
        if (weight <= 0.0) throw GraphEdgeError("Edge weights must be positive values")
    }

    override fun toString(): String {
        return "$start->$end"
    }
}

/*
 * This data class is returned for the Sequence generator for
 * the shortest path traversal.  The prior node is a nullable
 * because it includes the starting node with distance 0.
 */
data class PathEntry<T>(
    val node: GraphNode<T>,
    var distance: Double,
    var priorNode: GraphNode<T>?) {


    /*
     * We override the data class default toString so we have
     * something representative that makes sense and is good for testing.
     */
    override fun toString(): String {
        return "$node.name/$distance($priorNode?.name)"
    }
}

/*
 * Our graph design is using an adjacency list structure.  Each node
 * has a collection of edges, and each edge knows the start and end.
 *
 * We are assuming a directed graph.  Creating an undirected graph just
 * simply requires inserting one edge for each direction.
 */
class Graph<T>(){
    /*
     * We store the nodes in a key/value pair with the key being the "name"
     * of the node.  This allows us to easily find a particular node as the origin
     * of our shortest-path traversal
     */
    val nodes: MutableMap<T, GraphNode<T>> = mutableMapOf<T, GraphNode<T>>()

    /*
     * Our iterator is over nodes.  This trick is to convert the Map iterator
     * (which does key/value pairs) into an iterator that just has the value portion:
     * We turn the iterator into a Sequence, apply a map to the sequence (returning a
     * new sequence), and get the iterator from that sequence.
     */
    operator fun iterator() = nodes.iterator().asSequence().map{it.value}.iterator()

    /*
     * In inserting new nodes we check to ensure the key is not already in use,
     * but otherwise it accepts an arbitrary node name and creates an associated node.
     */
    fun insertNode(name: T) {
        if(name in nodes){
            throw GraphKeyError("No duplicate keys allowed")
        }
        nodes[name] = GraphNode(name)
    }

    /*
     * For edge creation we insert based on the name (rather than the nodes themselves).
     */
    fun insertEdge(start: T, end: T, weight: Double = 1.0){
        if(start !in nodes || end !in nodes){
            throw GraphKeyError("Failed to find node")
        }

        // We use the !! (non-null assertion) here.  We have checked
        // already that it is valid, but the compiler is unable to
        // assume this because possible concurrent access to a mutable
        // map means keys could vanish between checking and moving
        for(other in nodes[start]!!.outEdges){
            if (other.end == nodes[end]!!){
                throw GraphEdgeError("An edge already exists")
            }
        }
        val edge = GraphEdge(nodes[start]!!, nodes[end]!!, weight)
        nodes[start]!!.outEdges.add(edge)
        nodes[end]!!.inEdges.add(edge)
    }

    /*
     * This implements Dijkstra's shortest path algorithm as a Sequence.
     *
     * Since it is a sequence this means that each node is calculated on
     * demand, allowing someone using this implementation to iterate until
     * they find a destination they intend to reach.
     */
    fun shortestPath(start: T): Sequence<PathEntry<T>> {
        val workingSet = mutableMapOf<GraphNode<T>, PathEntry<T>>()
        if(start !in nodes) {
            throw GraphKeyError("Need to specify a node in the graph")
        }
        // We initialize all nodes to be an infinite distance away
        for ((_, node) in nodes){
            workingSet[node] = PathEntry(node, Double.POSITIVE_INFINITY, null)
        }
        // Except for the start node, that is.
        workingSet[nodes[start]]?.distance = 0.0

        // And here we do the lazy calculation of the shortest path on demand
        return sequence {
            // Now for the main loop, while there are still nodes in the working
            // set...
            while (workingSet.size > 0) {
                // We get the closest node still in the working set.
                var nextEntry: PathEntry<T>? = null
                for ((_, entry) in workingSet) {
                    // If nextEntry is null, the comparison will always be with infinity so
                    // we will take the new entry...
                    nextEntry = if (entry.distance <= (nextEntry?.distance ?: Double.POSITIVE_INFINITY)) {
                        entry
                    } else {
                        nextEntry
                    }
                }
                nextEntry as PathEntry<T>
                workingSet.remove(nextEntry.node)
                // If there are no more reachable nodes we will exit early.
                if (nextEntry.distance == Double.POSITIVE_INFINITY) {
                    return@sequence
                }
                // And now yield the result in the sequence
                yield(nextEntry)

                // This will now trigger if we want the next closest
                // node.  This is an advantage of the Sequence structure:
                // we are able to postpone these calculations until they are
                // actually needed.

                // So now we check each edge in current closest node...
                for (edge in nextEntry.node.outEdges) {
                    val distance = nextEntry.distance + edge.weight
                    // And if the distance is reduced by going from this node,
                    // we update that entry
                    if (edge.end in workingSet &&
                        workingSet[edge.end]!!.distance > distance) {
                        workingSet[edge.end]!!.distance = distance
                        workingSet[edge.end]!!.priorNode = nextEntry.node
                    }
                }
            }
        }
    }
}

fun <T>makeGraphNodes(vararg nodes:T) : Graph<T> {
    val graph = Graph<T>()
    for  (item in nodes){
        graph.insertNode(item)
    }
    return graph
}

